### PR TITLE
py action recognition demo: added i3d-rgb-tf model support

### PR DESCRIPTION
**Pull request by ivikhrev**
Added support of i3d-rgb-tf model for python action recognition demo. To distinguish models, it was necessary to add argument parameter `-model_type` with values `composite` for `encoder-decoder` models and `single` for i3d-rgb-tf model.

### CODE REVIEW

**Comment by vladimir-dudnik** [comment]
>  To distinguish models, it was necessary to add argument parameter -model_type with values composite for encoder-decoder models and single for i3d-rgb-tf model.

not clear, weld-porosity-detection-0001 is single model but listed under `composite` section. Is that intentional?

**Comment by ivikhrev** [comment]
> > To distinguish models, it was necessary to add argument parameter -model_type with values composite for encoder-decoder models and single for i3d-rgb-tf model.
> 
> not clear, weld-porosity-detection-0001 is single model but listed under `composite` section. Is that intentional?

 weld-porosity-detection-0001 is using Dummy Decoder in decoder step, so logically it refers to composite models with encoder-decoder pipeline. While i3d-rgb-tf doesn't need a decoder step in the pipeline, moreover this net accepts clip with duration 79 frames as whole input (others encoders accepts frames one by one), that's why I decided to distinguish it in different model type - `single`.

**Comment by ivikhrev** [comment]
Or, the third variant, remove Dummy Decoder at all, and move weld-porosity-detection-0001 to single models type.

**Comment by vladimir-dudnik** [comment]
probably the last (remove dummy decoder from weld-porosity case) is the best choice. @snosov1 don't you mind if we change way how weld-porosity-detection-0001 is integrated into action-recognition demo?

**Comment by vladimir-dudnik** [comment]
please update tests according to changes
action_recognition_demo.py: error: argument -at/--architecture_type: invalid choice: ' i3d-rgb' (choose from 'dummy-de', 'en-de', 'i3d-rgb')

**Review by Wovchena — CHANGES_REQUESTED** [review]

File: demos/action_recognition_demo/python/action_recognition_demo.py [hunk 1]
```diff
@@ -97,23 +98,28 @@ def main():
 
     encoder_xml = args.m_encoder
     encoder_bin = args.m_encoder.replace('.xml', '.bin')
-    encoder = IEModel(encoder_xml, encoder_bin, ie, encoder_target_device,
-                      num_requests=(3 if args.device == 'MYRIAD' else 1))
-
-    if args.m_decoder is not None:
+    model = []
```
Comment (eizamaliev) [inline demos/action_recognition_demo/python/action_recognition_demo.py:101]: I think it should be renamed to `models`

File: demos/action_recognition_demo/python/action_recognition_demo/result_renderer.py [hunk 2]
```diff
@@ -49,9 +50,13 @@ def __init__(self, no_show, presenter, output, limit, display_fps=False, display
...
-        self.meters['encoder'].update(timers['encoder'])
-        self.meters['decoder'].update(timers['decoder'])
-        return self.meters['encoder'].avg + self.meters['decoder'].avg
+        if self.architecture_type in ('en-de', 'dummy-de'):
+            self.meters['encoder'].update(timers['encoder'])
+            self.meters['decoder'].update(timers['decoder'])
+            return self.meters['encoder'].avg + self.meters['decoder'].avg
+        elif self.architecture_type == 'i3d-rgb':
+            self.meters['i3d-rgb-model'].update(timers['i3d-rgb-model'])
+            return  self.meters['i3d-rgb-model'].avg
```
[the remark this answers is not in the document]
Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo/result_renderer.py:59]: Function returns average values, It should be like this:
`inference_time += self.meters[key].avg`

File: demos/action_recognition_demo/python/action_recognition_demo.py [hunk 3]
```diff
@@ -97,23 +98,28 @@ def main():
...
+    model.append(IEModel(encoder_xml, encoder_bin, ie, encoder_target_device,
+                num_requests=(3 if args.device == 'MYRIAD' else 1)))
```
[the remark this answers is not in the document]
Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo.py]: Done, also changed name to models as @eizamaliev suggested

File: demos/action_recognition_demo/python/action_recognition_demo/models.py [hunk 4]
```diff
@@ -41,13 +41,13 @@ def adaptive_resize(frame, dst_size):
...
-    frame = adaptive_resize(frame, 224)
-    frame = center_crop(frame, (224, 224))
+def preprocess_frame(frame, size=224, crop_size=224, add_dimension=True):
+    frame = adaptive_resize(frame, size)
+    frame = center_crop(frame, (crop_size, crop_size))
 
     frame = frame.transpose((2, 0, 1))  # HWC -> CHW
-    frame = frame[np.newaxis, ...]  # add batch dimension
+    if add_dimension:
+        frame = frame[np.newaxis, ...]  # add batch dimension
```
[the remark this answers is not in the document]
Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo/models.py]: Changed

File: demos/action_recognition_demo/python/action_recognition_demo/result_renderer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo/result_renderer.py]: Done

File: demos/action_recognition_demo/python/action_recognition_demo/steps.py [hunk 5]
```diff
@@ -56,7 +88,6 @@ def process(self, item):
     def end(self):
         pass
 
-
```
[the remark this answers is not in the document]
Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo/steps.py]: Ok

File: demos/action_recognition_demo/python/action_recognition_demo.py [hunk 6]
```diff
@@ -97,23 +98,28 @@ def main():
...
+
+    if args.architecture_type == 'dummy-de':
+        model.append(DummyDecoder(num_requests=2))
+        seq_size = args.decoder_seq_size
+    elif args.architecture_type == 'en-de':
+        if args.m_decoder is None:
+            raise RuntimeError('No decoder for encoder-decoder model type (-m_de) provided: {}'.format(args.m_decoder))
```
[the remark this answers is not in the document]
Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo.py]: Ok

File: demos/action_recognition_demo/python/action_recognition_demo/steps.py [hunk 7]
```diff
@@ -26,17 +26,49 @@
...
     pipeline = AsyncPipeline()
     pipeline.add_step("Data", DataStep(capture), parallel=False)
-    pipeline.add_step("Encoder", EncoderStep(encoder), parallel=False)
-    pipeline.add_step("Decoder", DecoderStep(decoder, sequence_size=decoder_seq_size), parallel=False)
+
+    if model_type in ('en-de', 'dummy-de'):
+        pipeline.add_step("Encoder", EncoderStep(model[0]), parallel=False)
+        pipeline.add_step("Decoder", DecoderStep(model[1], sequence_size=seq_size), parallel=False)
+    elif model_type == 'i3d-rgb':
+        pipeline.add_step("I3DRGBModelStep", I3DRGBModelStep(model[0], seq_size, 256, 224), parallel = False)
```
[the remark this answers is not in the document]
Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo/steps.py]: Done

File: demos/action_recognition_demo/python/action_recognition_demo.py (hunk 1 above)
Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo.py]: Changed

**Review by Wovchena — APPROVED** [review]

File: demos/action_recognition_demo/python/README.md [hunk 8]
```diff
@@ -5,18 +5,21 @@
 This is the demo application for Action Recognition algorithm, which classifies actions that are being performed on input video.
 The following pre-trained models are delivered with the product:
 * `driver-action-recognition-adas-0002-encoder` + `driver-action-recognition-adas-0002-decoder`, which are models for driver monitoring scenario. They recognize actions like safe driving, talking to the phone and others
-* `action-recognition-0001-encoder` + `action-recognition-0001-decoder`, which are general-purpose action recognition (400 actions) models for Kinetics-400 dataset.
+* `action-recognition-0001-encoder` + `action-recognition-0001-decoder` and `i3d-rgb-tf`, which are general-purpose action recognition (400 actions) models for Kinetics-400 dataset.
 
-For more information about the pre-trained models, refer to the [model documentation](../../../models/intel/index.md).
+For more information about the pre-trained models, refer to the [intel models documentation](../../../models/intel/index.md) and [public models documentation](../../../models/public/index.md).
```
Comment (IRDonch) [inline demos/action_recognition_demo/python/README.md:10]:
```suggestion
For more information about the pre-trained models, refer to the [Intel model documentation](../../../models/intel/index.md) and [public model documentation](../../../models/public/index.md).
```

File: demos/action_recognition_demo/python/action_recognition_demo/steps.py [hunk 9]
```diff
@@ -26,18 +26,52 @@
...
     pipeline = AsyncPipeline()
     pipeline.add_step("Data", DataStep(capture), parallel=False)
-    pipeline.add_step("Encoder", EncoderStep(encoder), parallel=False)
-    pipeline.add_step("Decoder", DecoderStep(decoder, sequence_size=decoder_seq_size), parallel=False)
+
+    if model_type in ('en-de', 'dummy-de'):
+        pipeline.add_step("Encoder", EncoderStep(model[0]), parallel=False)
+        pipeline.add_step("Decoder", DecoderStep(model[1], sequence_size=seq_size), parallel=False)
+    elif model_type == 'i3d-rgb':
+        pipeline.add_step("I3DRGB", I3DRGBModelStep(model[0], seq_size, 256, 224), parallel = False)
```
Comment (IRDonch) [inline demos/action_recognition_demo/python/action_recognition_demo/steps.py:37]:
```suggestion
        pipeline.add_step("I3DRGB", I3DRGBModelStep(model[0], seq_size, 256, 224), parallel=False)
```

File: demos/action_recognition_demo/python/action_recognition_demo.py [hunk 10]
```diff
@@ -97,23 +98,27 @@ def main():
...
     encoder_xml = args.m_encoder
     encoder_bin = args.m_encoder.replace('.xml', '.bin')
-    encoder = IEModel(encoder_xml, encoder_bin, ie, encoder_target_device,
-                      num_requests=(3 if args.device == 'MYRIAD' else 1))
-
-    if args.m_decoder is not None:
+    models = [IEModel(encoder_xml, encoder_bin, ie, encoder_target_device,
+                num_requests=(3 if args.device == 'MYRIAD' else 1))]
+
+    if args.architecture_type == 'dummy-de':
```
Comment (IRDonch) [inline demos/action_recognition_demo/python/action_recognition_demo.py:104]: IMO, naming an architecture after what it _doesn't_ have is not the best strategy. I would suggest `en-mean` instead, which describes how the model works (you encode every frame and then you take the mean).

File: demos/action_recognition_demo/python/action_recognition_demo/steps.py [hunk 11]
```diff
@@ -26,18 +26,52 @@
...
+        self.async_model = AsyncWrapper(self.model, self.model.num_requests)
+
+    def process(self, frame):
+        preprocessed = preprocess_frame(frame, self.size, self.crop_size)
+        self.input_seq.append(preprocessed)
+        if len(self.input_seq) == self.sequence_size:
+            input_blob = np.array(self.input_seq)
+            input_blob = np.transpose(input_blob, (1, 0, 2, 3))
+            input_blob = np.expand_dims(input_blob, axis=0)
+            output, next_frame = self.async_model.infer(input_blob, frame)
```
Comment (IRDonch) [inline demos/action_recognition_demo/python/action_recognition_demo/steps.py:65]: Doesn't this model take RGB input? Shouldn't the input be converted to RGB at some point?

**Review by IRDonch — CHANGES_REQUESTED** [review]

File: demos/action_recognition_demo/python/README.md (hunk 8 above)
Comment (ivikhrev) [reply demos/action_recognition_demo/python/README.md]: Ok

File: demos/action_recognition_demo/python/action_recognition_demo/steps.py (hunk 9 above)
Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo/steps.py]: Ok

File: demos/action_recognition_demo/python/action_recognition_demo.py (hunk 10 above)
Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo.py:104]: Yes, I agree, `en-mean` would be better, thanks.

File: demos/action_recognition_demo/python/action_recognition_demo/steps.py (hunk 11 above)
Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo/steps.py:65]: Yeah, good point, I thought, don't know why, I did it somewhere with transpose function. It's nice that you've noticed it. But, what surprise me, this model works fine on tested videos and images even with BGR channel order, without noticable changes in output  (even without changes at all, I would say). That's why I haven't noticed it myself. Thx!

Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo/steps.py:66]: Fixed

File: demos/action_recognition_demo/python/action_recognition_demo.py (hunk 10 above)
Comment (ivikhrev) [reply demos/action_recognition_demo/python/action_recognition_demo.py]: Updated

**Review by IRDonch — APPROVED** [review]

**Pull request merged**